### PR TITLE
Modular graphnav

### PR DESCRIPTION
**Pull request by heuristicus**

### CODE REVIEW

File: spot_wrapper/spot_graph_nav.py [hunk 1]
```diff
@@ -0,0 +1,926 @@
...
+        ).to_proto()
+        # Create an empty instance for initial localization since we are asking it to localize
+        # based on the nearest fiducial.
+        localization = nav_pb2.Localization()
+        self._graph_nav_client.set_localization(
+            initial_guess_localization=localization,
+            ko_tform_body=current_odom_tform_body,
+        )
+
+    def set_initial_localization_waypoint(self, *args):
```
Comment (bhung-bdai) [inline spot_wrapper/spot_graph_nav.py:298]: Same duplication as above

File: spot_wrapper/spot_graph_nav.py [hunk 2]
```diff
@@ -0,0 +1,926 @@
...
+        state = self._graph_nav_client.get_localization_state()
+        self._logger.info(f"Got localization: \n{str(state.localization)}")
+        odom_tform_body = get_odom_tform_body(
+            state.robot_kinematics.transforms_snapshot
+        )
+        self._logger.info(
+            f"Got robot state in kinematic odometry frame: \n{str(odom_tform_body)}"
+        )
+
+    def set_initial_localization_fiducial(self, *args):
```
Comment (bhung-bdai) [inline spot_wrapper/spot_graph_nav.py:284]: Same duplication as above.

File: spot_wrapper/spot_graph_nav.py [hunk 3]
```diff
@@ -0,0 +1,926 @@
...
+                f"Got an error during downloading graph and snapshots from the robot: {e}",
+            )
+
+    ## Copied from https://github.com/boston-dynamics/spot-sdk/blob/master/python/examples/graph_nav_command_line/recording_command_line.py and https://github.com/boston-dynamics/spot-sdk/blob/master/python/examples/graph_nav_command_line/graph_nav_command_line.py with minor modifications
+    # Copyright (c) 2020 Boston Dynamics, Inc.  All rights reserved.
+    #
+    # Downloading, reproducing, distributing or otherwise using the SDK Software
+    # is subject to the terms and conditions of the Boston Dynamics Software
+    # Development Kit License (20191101-BDSDK-SL).
+    def _get_localization_state(self, *args):
```
Comment (bhung-bdai) [inline spot_wrapper/spot_graph_nav.py:273]: This function appears to be duplicated in `wrapper.py`

File: spot_wrapper/spot_graph_nav.py [hunk 4]
```diff
@@ -0,0 +1,926 @@
...
+            )
+
+    def _write_bytes(self, filepath: str, filename: str, data):
+        """Write data to a file."""
+        os.makedirs(filepath, exist_ok=True)
+        with open(filepath + filename, "wb+") as f:
+            f.write(data)
+            f.close()
+
+    def _list_graph_waypoint_and_edge_ids(self, *args):
```
Comment (bhung-bdai) [inline spot_wrapper/spot_graph_nav.py:426]: Almost exactly the same as the function in wrapper.py

File: spot_wrapper/spot_graph_nav.py [hunk 5]
```diff
@@ -0,0 +1,926 @@
...
+        # any navigation commands.
+        localization_state = self._graph_nav_client.get_localization_state()
+        if not localization_state.localization.waypoint_id:
+            # The robot is not localized to the newly uploaded graph.
+            self._logger.info(
+                "Upload complete! The robot is currently not localized to the map; please localize the robot using a "
+                "fiducial before attempting a navigation command."
+            )
+
+    def _navigate_to(self, waypoint_id: str) -> typing.Tuple[bool, str]:
```
Comment (bhung-bdai) [inline spot_wrapper/spot_graph_nav.py:534]: Again, similar duplication in wrapper.py

File: spot_wrapper/spot_graph_nav.py [hunk 6]
```diff
@@ -0,0 +1,926 @@
...
+        Args:
+            filepath (str) : Path of file where data will be written.
+            data (bytes) : Bytes of data"""
+        directory = os.path.dirname(filepath)
+        os.makedirs(directory, exist_ok=True)
+        with open(filepath, "wb+") as f:
+            f.write(data)
+            f.close()
+
+    def _download_graph_and_snapshots(
```
Comment (bhung-bdai) [inline spot_wrapper/spot_graph_nav.py:162]: Same duplication in wrapper.py

File: spot_wrapper/spot_graph_nav.py [hunk 7]
```diff
@@ -0,0 +1,926 @@
...
+                    "Failed to download edge snapshot: %s", edge.snapshot_id
+                )
+                continue
+            self._write_bytes_while_download(
+                os.path.join(download_path, "edge_snapshots", edge.snapshot_id),
+                edge_snapshot.SerializeToString(),
+            )
+        return True, "Success"
+
+    def clear_graph(self) -> typing.Tuple[bool, str]:
```
Comment (bhung-bdai) [inline spot_wrapper/spot_graph_nav.py:209]: Similar function without the extra nested function call in `wrapper.py`

**Comment by heuristicus** [comment]
This was just a reference for me, please use https://github.com/bdaiinstitute/spot_wrapper/pull/44 for comments.

File: spot_wrapper/spot_graph_nav.py [hunk 8]
```diff
@@ -0,0 +1,926 @@
...
+        try:
+            self._clear_graph()
+            return True, "Success"
+        except Exception as e:
+            return (
+                False,
+                f"Got an error while clearing a graph and snanshots in a robot: {e}",
+            )
+
+    def upload_graph(self, upload_path: str) -> typing.Tuple[bool, str]:
```
Comment (bhung-bdai) [inline spot_wrapper/spot_graph_nav.py:222]: Duplication in `wrapper.py`

**Pull request closed without merging**